pkg/controller/certificatesigningrequest/approving_controller.go

```diff
@@ -0,0 +1,177 @@
+	c := &CSRApprovingController{
+		client:          client,
+		csrLister:       csrInformer.Lister(),
+		csrListerSynced: csrInformer.Informer().HasSynced,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "certificateSigningRequest"),
+		approvers: []approver{
+			&ipsecCSRApprover{
+				client: client,
+			},
+		},
+	}
+	csrInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueueCertificateSigningRequest,
+		},
+		resyncPeriod,
+	)
+	return c
+}
+
+// Run begins watching and syncing of the CSRApprovingController.
+func (c *CSRApprovingController) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.InfoS("Starting " + approvingControllerName)
+	defer klog.InfoS("Shutting down " + approvingControllerName)
+
+	cacheSyncs := []cache.InformerSynced{c.csrListerSynced}
+	if !cache.WaitForNamedCacheSync(approvingControllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *CSRApprovingController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *CSRApprovingController) enqueueCertificateSigningRequest(obj interface{}) {
+	csr, ok := obj.(*certificatesv1.CertificateSigningRequest)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "object", obj)
+			return
+		}
+		csr, ok = deletedState.Obj.(*certificatesv1.CertificateSigningRequest)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-CertificateSigningRequest object", "object", deletedState.Obj)
+			return
+		}
+	}
+	c.queue.Add(csr.Name)
+}
+
+func (c *CSRApprovingController) syncCSR(key string) error {
+	startTime := time.Now()
+	defer func() {
+		d := time.Since(startTime)
+		klog.V(2).InfoS("Finished syncing CertificateSigningRequest", "name", key, "duration", d)
+	}()
+
+	csr, err := c.csrLister.Get(key)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	// The spec will not be updated by antrea-agent once it is approved or denied.
+	if approved, denied := getCertApprovalCondition(&csr.Status); approved || denied {
+		return nil
+	}
+	for _, a := range c.approvers {
+		if a.recognize(csr) {
+			approved, err := a.verify(csr)
+			if err != nil {
+				return err
+			}
+			if approved {
+				toUpdate := csr.DeepCopy()
+				appendApprovalCondition(toUpdate, fmt.Sprintf("Automatically approved by %s", a.name()))
+				_, err = c.client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(), toUpdate.Name, toUpdate, metav1.UpdateOptions{})
+				if err != nil {
+					return fmt.Errorf("error updating approval for csr: %v", err)
+				}
+				return nil
+			}
```

Comment:
should it deny the CSR if recongnized but not approved because of non transientError?

Reply:
I am not sure whether it is necessary to deny the CSR as K8s built-in CSR controllers never do so. But I am ok to add the deny logic back.

Reply:
OK, if other signers don't deny them, I'm ok with using same stragegy